ref.c

```diff
@@ -0,0 +1,406 @@
+            fn += strlen(fn);
+            cp++;
+        }
+        else if (*cp >= '0' && *cp <= '9')
+        {
+            char *endp;
+            long l;
+
+            l = strtol(cp, &endp, 10);
+            l = MIN(l, strlen(fn));
+            if (*endp == 's')
+            {
+                strncpy(path, fn, l);
+                path += l;
+                fn += l;
+                *path = 0;
+                cp = endp + 1;
+            }
+            else
+            {
+                *path++ = '%';
+                *path++ = *cp++;
+            }
+        }
+        else
+        {
+            *path++ = '%';
+            *path++ = *cp++;
+        }
+        dir = cp;
+    }
+    strcpy(path, dir);
+    path += strlen(dir);
+    if (*fn && path[-1] != '/')
+        *path++ = '/';
+    strcpy(path, fn);
+}
+
+/*
+ * Return an integer representation of pthread_self().
+ */
+static unsigned get_int_threadid()
+{
+    pthread_t pt = pthread_self();
+    unsigned char *s = (unsigned char *)&pt;
+    size_t i;
+    unsigned h = 0;
+    for (i = 0; i < sizeof(pthread_t); i++)
+        h = (h << 5) - h + s[i];
+    return h;
+}
+
+/*
+ * Make the directory containing path and any prefix directories.
+ */
+void mkdir_prefix(char *path, int mode)
+{
+    char *cp = strrchr(path, '/');
+    if (!cp)
+        return;
+
+    *cp = 0;
+    if (is_directory(path))
+    {
+        *cp = '/';
+        return;
+    }
+
+    if (mkdir(path, mode) == 0)
+    {
+        chmod(path, mode);
+        *cp = '/';
+        return;
+    }
+
+    mkdir_prefix(path, mode);
+    mkdir(path, mode);
+    chmod(path, mode);
+    *cp = '/';
+}
+
+int ref_close(hFILE* hf){
+    hFILE_ref *hfref = (hFILE_ref*)hf;
+    if(hfref->file_name) free(hfref->file_name);
+
+    if (hfref->mf) {
+        if (mfclose(hfref->mf) != 0) return -1;
+
+        // The buffer is freed in mfclose, but we need something for
+        // hclose to free (the caller of this), or it will error
+        ((hFILE_wrapper*)hf)->innerhf->buffer = malloc(1);
+    }
+
+    return 0;
+}
+
+static const struct hFILE_backend ref_backend =
+{
+    NULL, NULL, NULL, NULL, ref_close
+};
+
+/*
+ * Modifies the base_hFILE parameter to contain the length and filename information
+ * as a hFILE_ref
+*/
+static hFILE* open_hfile_ref(hFILE* base_hFILE, char* file_name, int length, mFILE* mf, int is_mem_hfile){
+    hFILE_ref* hfref = (hFILE_ref *) hfile_init_wrapper(sizeof (hFILE_ref), base_hFILE);
+
+    hfref->file_name = file_name;
+    hfref->length = length;
+    hfref->mf = mf;
+    hfref->base.base.backend = &ref_backend;
+    hfref->is_mem_hfile = is_mem_hfile;
+
+    return &hfref->base.base;
+}
+
+#define READ_CHUNK_SIZE 8192
+
+/*
+ * Puts the char* into a hfile
+*/
+hFILE* mFILE_to_hFILE_ref(mFILE* mf){
+    hFILE *hf = NULL;
+    size_t sz;
+    char* seq;
+    seq = mfsteal(mf, &sz);
+
+    if(!seq){
+        seq = mf->data;
+    }
+    else{
+        mf = NULL;
+    }
+
```

Comment:
I'm unsure about the above.

`mfsteal` returns the buffer and size from the mFILE and then destroys (frees) the mFILE.  It's designed to turn an mFILE into a C buffer basically.  Thus after the `mfsteal` call `mf` is now undefined and we cannot dereference it and nor should we pass it to `open_hfile_ref`.

Reply:
Hmm looking further I see how it works.  It only uses mf if mfsteal failed, otherwise it sets mf to NULL.  mfsteal can only fail if mfdetach fails, which in turn only fails if we're using mmap.

I think the logic in this function is therefore correct (sorry), but I'm wondering if it's a symptom of my mfsteal being too complex in error handling.

At the very least we probably should do something like this to get to a consistent point whether or not the steal fails:

```
if (!seq) {
    seq = mf->data;
    mf->data = NULL;
    mfdestroy(mf);
}
```

and remove the mf parameter from `open_file_href`.  This all ties up with the ref_close comments though too, as that's where `href->mf` is used.

Reply:
Sorry for spamming.  I've finally figured out what's happening here and I feel guilty for commenting on what turns out to be entirely my own demons!

So, just incase anyone else who is reading this cares...

The original code in cram_populate_ref was:

```
    /* Otherwise search full REF_PATH; slower as loads entire file */
    if ((mf = open_path_mfile(tag->str+3, ref_path, NULL))) {
        size_t sz;
        r->seq = mfsteal(mf, &sz);
        if (r->seq) {
            r->mf = NULL;
        } else {
            // keep mf around as we couldn't detach                                                              
            r->seq = mf->data;
            r->mf = mf;
        }
        r->length = sz;
        r->is_md5 = 1;
    } else {
```

(The comment would probably have been useful to keep, albeit with explanation that mfdetach won't free the data if we used mmap.)

Basically if the `open_path_mfile` used mmap on an md5 file then we need to keep the file descriptor open as we need it for munmap later on.  Otherwise once it's open we can just steal the pointer can close it.  This would have been the original way it was done, with mmap appearing later on.  With hindsight it probably should have kept mf valid in all scenarios, for simplicity, and never attempted the mfsteal part at all.   We probably needed a "mf_close_if_able" function or similar that closed the file descriptor if not mmaped so it's no longer consuming resources, but keep the rest of the mFILE struct present.

Hence hFILE has now taken the role as replacement for `ref_entry *r` and `r->mf`.  It's the same code as my original, just with different structure names.  Sorry Thomas. :)

I think what all of this points towards is that mFILE is overkill for what we need here.  It was designed as a transparent drop in for stdio FILE I/O so it supports read/write, seeking, update, etc.  We really just need a load-or-mmap function, which is what the hfile mem layer provides (minus the missing mmap bit).

PS. My head hurts!